packages/core/saved-objects/core-saved-objects-base-server-internal/src/model_version/model_version_observer/model_version_observer.ts

```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import * as Rx from 'rxjs';
+import deepEqual from 'fast-deep-equal';
+import { errors } from '@elastic/elasticsearch';
+import type { IndicesGetMappingResponse } from '@elastic/elasticsearch/lib/api/types';
+import type { ElasticsearchClient } from '@kbn/core-elasticsearch-server';
+import type { Logger } from '@kbn/logging';
+import pRetry from 'p-retry';
+
+import type { IndexMapping } from '../../mappings';
+import { getModelVersionsFromMappings } from '../model_version_from_mappings';
+import type { ModelVersionMap } from '../version_map';
+
+export interface Dependencies {
+  client: ElasticsearchClient;
+  logger: Logger;
+  /**
+   * How long to wait before polling again.
+   *
+   * @note this does not gaurantee polling will occur every `pollInterval` ms, but rather
+   *       after a poll has occurred we will wait at least `pollInterval` ms before polling again.
+   */
+  pollInterval?: number;
+}
+
+const POLL_INTERVAL_MS = 30_000;
+/** TODO: Is this the correct pattern? In this way we will also be matching indices after "the split" */
+const KIBANA_SYSTEM_INDICES_PATTERN = '.kibana_*';
+
+async function robustlyFetchMappings(
+  client: ElasticsearchClient
+): Promise<IndicesGetMappingResponse> {
+  const run = async () => {
+    try {
+      return await client.indices.getMapping({
+        index: KIBANA_SYSTEM_INDICES_PATTERN,
+        expand_wildcards: 'hidden',
+      });
+    } catch (e) {
+      if (
+        e instanceof errors.ResponseError &&
```

Comment:
What's the risk of an intermittent proxy response? Should we also be checking the headers?